tools/user_migration/migrate_user.py

```diff
@@ -0,0 +1,288 @@
+        SELECT
+            email, generation, keys_changed_at, client_state
+        FROM users
+            WHERE uid = {uid}
+    """.format(uid=user)
+    cursor = databases['mysql'].cursor()
+    cursor.execute(sql)
+    (email, generation, keys_changed_at, client_state) = cursor.next()
+    fxa_uid = email.split('@')[0]
+    fxa_kid = format_key_id(
+        keys_changed_at or generation,
+        bytes.fromhex(client_state),
+    )
+    cursor.close()
+    return (fxa_kid, fxa_uid)
+
+
+def dumper(columns, values):
+    """verbose column and data dumper. """
+    for row in values:
+        print("---\n")
+        for i in range(0, len(columns)):
+            print("{} => {}".format(columns[i], row[i]))
+    print("===\n")
+
+
+def move_user(databases, user, args):
+    """copy user info from original storage to new storage."""
+    # bso column mapping:
+    # id => bso_id
+    # collection => collection_id
+    # sortindex => sortindex
+    # modified => modified
+    # payload => payload
+    # payload_size => NONE
+    # ttl => expiry
+
+    # user collections require a unique key.
+    unique_key_filter = []
+
+    uc_columns = (
+        'fxa_kid',
+        'fxa_uid',
+        'collection_id',
+        'modified',
+    )
+    bso_columns = (
+            'collection_id',
+            'fxa_kid',
+            'fxa_uid',
+            'bso_id',
+            'expiry',
+            'modified',
+            'payload',
+            'sortindex',
+    )
+
+    # Genereate the Spanner Keys we'll need.
+    (fxa_kid, fxa_uid) = get_fxa_id(databases, user)
+
+    # Fetch the BSO data from the original storage.
+    sql = """
+    SELECT collection, id, ttl, modified, payload, sortindex
+    FROM bso
+    WHERE userid = {}
+    ORDER BY modified DESC"""
+
+    count = 0
+    with databases['spanner'].batch() as batch:
+        cursor = databases['mysql'].cursor()
+        try:
+            cursor.execute(sql.format(user))
+            print("Processing... {} -> {}:{}".format(
+                user, fxa_uid, fxa_kid))
+            for (cid, bid, exp, mod, pay, sid) in cursor:
+                # columns from sync_schema3
+                mod_v = datetime.utcfromtimestamp(mod/1000)
+                exp_v = datetime.utcfromtimestamp(exp)
+                # User_Collection can only have unique values. Filter
+                # non-unique keys and take the most recent modified
+                # time. The join could be anything.
+                uc_key = "{}_{}_{}".format(fxa_uid, fxa_kid, cid)
+                if uc_key not in unique_key_filter:
+                    unique_key_filter.append(uc_key)
+                    uc_values = [(
+                        fxa_kid,
+                        fxa_uid,
+                        cid,
+                        mod_v,
+                    )]
+                    if args.verbose:
+                        print("### uc:")
+                        dumper(uc_columns, uc_values)
+                    batch.insert(
```

Comment:
we need some additional work here somewhat similar to what this block is doing: non std entries in the `collections` table should be lazily added there

Reply:
Ah, good point.

Reply:
Wait, looking at the code, I'm lazily adding `user_collections`, but maybe I was presuming that `collections` is fairly static. Do I need to add unique user collections to that table as well?

Reply:
Nope, this part is correct. Users get a `user_collections` entry for each collection they have `bsos` for.

And every unique collection gets its own entry in the top level `collections` table, storing its id and name shared between all users. So users end up creating arbitrary collections when storing bsos with a previously unknown collection for the firs time.

Reply:
🐉🐉🐉🐉🐉🐉🐉🐉
... 
Ok, so there may be a problem here. Since `user_collections` are "unique" to a node, there's the potential that two different nodes have different values for "unique", this could lead to collisions or duplication. 

Need to research this a bit to determine the impacts or how we want to handle this. 

🐉🐉🐉🐉🐉🐉🐉🐉

Reply:
Looking into the problem:

`collection` is the only place where a given `collection_id` is mapped to a `collection_name`. `collection_name` is used as one of the primary locking mechanisms. Most users only have well known collections, so it should be possible to return a well known collection id. The problem is that some users may have "custom" collections which have a different value. 

Option 1:
Spanner makes modifications of table indexes problematic. We *could* modify user_collections to include "name" as an index field, but we would have to either replicate the data in the existing user_collections to a new table, or just create a new table that contains
`fxa_uid`, `collection_id`, `collection_name` (all indices) and use that as the fail over lookup table for collection_id if it's not one of the well-known ones.

Option 2:
Modify user_collection tables to include the name field and do a secondary index lookup on that if no well-known collections found.

There may be other options as well, but those are the two I've thought of tonight.

No idea how this might impact adding new users 2020-01-13

Reply:
Another option:

Client appears to only provide collection by name, which is a `String`. The migration script could load the known `user_collections` into a local cache, and remap the collection IDs per user. This should require the least amount of db modification, but potentially could impact the speed of migration.